cub/block/block_run_length_decode.cuh

```diff
@@ -0,0 +1,435 @@
+   * used to retrieve the remaining run-length decoded items.
+   *
+   * @param from_decoded_offset If invoked with from_decoded_offset that is larger than total_decoded_size results in
+   * undefined behavior.
+   */
+  template <typename UserRelativeOffsetT>
+  __device__ __forceinline__ void RunLengthDecode(UniqueItemT (&decoded_items)[DECODED_ITEMS_PER_THREAD],
+                                                  UserRelativeOffsetT (&item_offsets)[DECODED_ITEMS_PER_THREAD],
+                                                  int32_t from_decoded_offset = 0)
+  {
+    // Populate run-length decode buffer with continuation-of-current-run items
+    InitDecodeBuffer(temp_storage.runs.decode_buffer);
+
+    // Ensure that the continuation-of-current-run items have been written to shared memory
+    __syncthreads();
+
+    // Scatter the beginning of each run into the run-length decode buffer
+    RunOffsetT run_offset = linear_tid;
+#pragma unroll
+    for (uint32_t i = 0; i < RUNS_PER_THREAD; i++)
+    {
+      // Fetch this thread's run and subsequently check whether it falls into this pass' run-length decode buffer
+      int32_t run_target_idx      = temp_storage.runs.run_offsets[run_offset];
+      int32_t next_run_target_idx = temp_storage.runs.run_offsets[run_offset + 1];
+      int32_t run_length          = next_run_target_idx - run_target_idx;
+
+      // Whether this run begins before the from_decoded_offset but has trailing items that still fall within the
+      // beginning of this buffer
+      bool is_a_trailing_run = (run_target_idx < from_decoded_offset) && (next_run_target_idx > from_decoded_offset);
+
+      // In particular for the trailing run, we want to scatter into index 0 (not a negative scatter index)
+      int32_t scatter_target_idx = max(0, run_target_idx - from_decoded_offset);
+      int32_t relative_offset    = is_a_trailing_run ? from_decoded_offset - run_target_idx : 0;
+
+      // Only write this beginning-of-run if the run length falls within the decoded buffer OR
+      // the run is a trailing run.
+      if (run_length > 0 && (run_target_idx >= from_decoded_offset || is_a_trailing_run) &&
+          scatter_target_idx < BLOCK_DECODED_ITEMS)
+      {
+        WriteBeginningOfRun(temp_storage.runs.decode_buffer,
+                            scatter_target_idx,
+                            temp_storage.runs.unique_items[run_offset],
+                            relative_offset);
+      }
+      run_offset += BLOCK_THREADS;
+    }
+
+    // Ensure that the beginning of all relevant runs has been written
+    __syncthreads();
+
+    // Read in blocked-arrangement in preparation for the run-length decoding prefix scan
+    ZippedRunLengthDecode scan_items[DECODED_ITEMS_PER_THREAD];
+    LoadFromDecodeBuffer(scan_items, temp_storage.runs.decode_buffer);
+
+    // Ensure the decoded_buffer can be repurposed for the scan's temporary storage
+    __syncthreads();
```

Comment:
For some reason, we use `CTA_SYNC()` macro instead of `__syncthraeds()` everywhere in CUB 😄

Reply:
Changed it to `CTA_SYNC()´ 👍